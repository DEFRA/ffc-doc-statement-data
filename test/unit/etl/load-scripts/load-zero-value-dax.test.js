const db = require('../../../../app/data')
const { loadZeroValueDax } = require('../../../../app/etl/load-scripts/load-zero-value-dax')

jest.mock('../../../../app/data', () => ({
  sequelize: {
    query: jest.fn()
  }
}))

describe('loadZeroValueDax', () => {
  const startDate = '2023-01-01'
  const transaction = {}

  beforeEach(() => {
    db.sequelize.query.mockClear()
  })

  test('should call sequelize.query with correct SQL and parameters', async () => {
    await loadZeroValueDax(startDate, transaction)

    expect(db.sequelize.query).toHaveBeenCalledTimes(1)
    expect(db.sequelize.query.mock.calls[0][0]).toMatchSnapshot()
    expect(db.sequelize.query.mock.calls[0][1]).toEqual({
      replacements: { startDate },
      raw: true,
      transaction
    })
  })

  test('should handle errors thrown by sequelize.query', async () => {
    db.sequelize.query.mockRejectedValue(new Error('Query failed'))

    await expect(loadZeroValueDax(startDate, transaction)).rejects.toThrow('Query failed')
  })
})
