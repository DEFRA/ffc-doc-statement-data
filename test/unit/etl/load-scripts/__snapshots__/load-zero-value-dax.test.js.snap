// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadZeroValueDax should call sequelize.query with correct SQL and parameters 1`] = `
"
    WITH unique_rows AS (
      SELECT DISTINCT ON (T."paymentRef", T."calculationId")
        T."paymentRef" AS "paymentReference",
        T."calculationId" AS "calculationId",
        T.quarter AS "paymentPeriod",
        T."totalAmount" AS "paymentAmount",
        T.transdate AS "transactionDate"
      FROM public."etlIntermTotalZeroValues" T
      LEFT JOIN public."delinkedCalculation" D ON T."calculationId" = D."calculationId"
      WHERE T."etlInsertedDt" > :startDate
        AND D."calculationId" IS NULL
      ORDER BY T."paymentRef", T."calculationId", T."etlInsertedDt" DESC
    )
    INSERT INTO public."zeroValueDax" (
      "paymentReference", "calculationId", "paymentPeriod",
      "paymentAmount", "transactionDate"
    )
    SELECT 
      "paymentReference", "calculationId", "paymentPeriod",
      "paymentAmount", "transactionDate"
    FROM unique_rows
;
  "
`;
